### PR TITLE
feature: OIDC provider scope API

### PR DESCRIPTION
**Pull request by fairclothjm**
This PR depends on https://github.com/hashicorp/vault/pull/12198

## Description
The scope API allows users to define custom scopes and associated claims to be made available on servers.

Acceptance Criteria:
- A Vault user can perform operations using the scope API as specified in the RFC.

## Manual Test

```
$ TOKEN_TEMPLATE=$(cat << EOF
{
    "groups": {{identity.entity.groups.names}}
}
EOF
)

$ vault write identity/oidc/scope/my-scope template="$(echo $TOKEN_TEMPLATE | base64 -)"
Success! Data written to: identity/oidc/scope/my-scope

$ vault list identity/oidc/scope
Keys
----
my-scope

$ vault read identity/oidc/scope/my-scope
Key            Value
---            -----
description    n/a
template       eyAiZ3JvdXBzIjoge3tpZGVudGl0eS5lbnRpdHkuZ3JvdXBzLm5hbWVzfX0gfQo=

$ vault write identity/oidc/scope/my-scope description="my description 123"
Success! Data written to: identity/oidc/scope/my-scope

$ vault read identity/oidc/scope/my-scope
Key            Value
---            -----
description    my description 123
template       eyAiZ3JvdXBzIjoge3tpZGVudGl0eS5lbnRpdHkuZ3JvdXBzLm5hbWVzfX0gfQo=

$ vault delete identity/oidc/scope/my-scope
Success! Data deleted (if it existed) at: identity/oidc/scope/my-scope

$ vault list identity/oidc/scope
No value found at identity/oidc/scope

$ vault read identity/oidc/scope/my-scope
No value found at identity/oidc/scope/my-scope

```

This PR depends on https://github.com/hashicorp/vault/pull/12198

### CODE REVIEW

File: vault/identity_store_oidc_provider.go
Comment (austingebauer) [inline vault/identity_store_oidc_provider.go:248]: We'll want to attempt to base64 decode and validate the template similar to how they're treated for identity token roles: [identity_store_oidc.go#L955-L984](https://github.com/hashicorp/vault/blob/main/vault/identity_store_oidc.go#L955-L984). There are also some reserved claims that we'll want to prevent the template from overwriting.

File: vault/identity_store_oidc_provider.go
Comment (austingebauer) [inline vault/identity_store_oidc_provider.go:328]: Leaving a note that we'll likely need to prevent deletion of scopes that are being referenced by providers. We can add that check with the provider PR (similar to the client/assignment situation). Won't block this PR 👍

File: vault/identity_store_oidc_provider.go
Comment (austingebauer) [inline vault/identity_store_oidc_provider.go:256]: I see that this is exactly similar to the template check for identity token roles. I'm wondering if using [ValidityCheckOnly](https://github.com/hashicorp/vault/blob/main/sdk/helper/identitytpl/templating.go#L29) makes sense here. Otherwise, [templating](https://github.com/hashicorp/vault/blob/main/sdk/helper/identitytpl/templating.go#L151) is performed. I'll look at this a bit closer.

I'm also curious about the `// namespace?` comment. Looking into that as well.

File: vault/identity_store_oidc_provider.go
Comment (austingebauer) [inline vault/identity_store_oidc_provider.go:86]: nit: We could improve on the description here a bit. See the description for [identity_store_oidc.go#L249](https://github.com/hashicorp/vault/blob/main/vault/identity_store_oidc.go#L249).

Maybe something like:
> The template string to use for the scope. This may be in string-ified JSON or base64 format.

**Review by austingebauer — COMMENTED** [review]
This is looking good! Just a few comments.

File: vault/identity_store_oidc_provider.go
Comment (fairclothjm) [reply vault/identity_store_oidc_provider.go:328]: CC @vinay-gopalan

Comment (fairclothjm) [reply vault/identity_store_oidc_provider.go:328]: Just realized that Vinay's PR is branched from the `oidc-provider-assignment-api` branch which does not yet have the Scope code. Maybe this can be addressed in a separate issue after all is merged?

Comment (vinay-gopalan) [reply vault/identity_store_oidc_provider.go:328]: If need be I can merge the this branch into the `oidc-provider-api` branch to get access to the Scope code. Things should be sorted as long as we merge the assignment and scope API PRs before the Provider API, I think

File: vault/identity_store_oidc_provider.go
Comment (austingebauer) [inline vault/identity_store_oidc_provider.go:222]: We should return an error if a user tries to create the `openid` scope. The `openid` scope is reserved and maps to a specific set of reserved claims.

**Review by austingebauer — APPROVED** [review]
LGTM

**Pull request merged**
